### PR TITLE
Save & Load

### LINKED ISSUE #5: Jagged lines are created when using touch on a mobile device
I don't remember this being a problem when I tried https://jsfiddle.net/jnm236/d67kxw1q/1/show on my Android.

Screenshot_20201120-010309_Brave

### PR DESCRIPTION
**Pull request by Daynvheur**
Saving and loading now available. Also fixes #5.

Check http://localhost:1234/?&pt=448;179&pt=448;180&pt=448;182&pt=442;194&pt=436;210&pt=434;228&pt=436;246&pt=441;258&pt=449;267&pt=460;278&pt=473;288&pt=490;299&pt=512;312&pt=537;324&pt=571;336&pt=605;341&pt=642;344&pt=679;345&pt=722;343&pt=765;339&pt=807;332&pt=854;321&pt=890;312&pt=922;301&pt=938;291&pt=943;283&pt=943;275&pt=942;265&pt=938;254&pt=933;245&pt=929;239&pt=921;233&pt=913;229&pt=903;227&pt=887;229&pt=869;237&pt=843;249&pt=808;268&pt=780;288&pt=757;310&pt=735;336&pt=720;359&pt=709;379&pt=702;399&pt=700;414&pt=699;426&pt=698;439&pt=699;450&pt=700;462&pt=703;475&pt=709;489&pt=721;507&pt=734;523&pt=753;541&pt=775;554&pt=802;567&pt=836;576&pt=867;579&pt=893;576&pt=910;568&pt=921;554&pt=924;539&pt=923;519&pt=916;503&pt=905;489&pt=881;474&pt=848;466&pt=806;463&pt=757;462&pt=694;463&pt=640;468&pt=590;473&pt=545;480&pt=500;487&pt=461;494&pt=425;500&pt=393;512&pt=372;524&pt=356;536&pt=344;550&pt=335;564&pt=328;577&pt=322;590&pt=319;598&pt=316;605&pt=316;613&pt=319;622&pt=326;631&pt=337;642&pt=355;648&pt=383;649&pt=420;646&pt=454;636&pt=485;619&pt=508;599&pt=522;576&pt=528;557&pt=532;541&pt=538;526&pt=544;516&pt=554;506&pt=566;491&pt=577;473&pt=589;451&pt=602;426&pt=613;405&pt=623;384&pt=633;359&pt=640;334&pt=645;307&pt=647;276&pt=645;247&pt=640;219&pt=631;194&pt=621;179&pt=611;166&pt=599;156&pt=585;146&pt=568;138&pt=549;131&pt=523;126&pt=489;123&pt=465;123&pt=453;125&pt=452;126&pt=448;179 out!

### CODE REVIEW

**Comment by jnm2** [comment]
Ooh, very nice! What do you think about using differential encoding, so that each point is the X difference and Y difference from the previous point? So for example it would start:

http://localhost:1234/?&pt=448;179&pt=0;1&pt=0;2&pt=-6;6&pt=6;16

**Comment by Daynvheur** [comment]
I don't have a strong opinion on either absolute or ref points.
I may have 2 little 'pro' favouring abs:
- If tampering with the values (deleting one point, then looking for reinserting afterwards), this will be simpler to just check asbolute values vs references to previous points ; but that'll end up totally unreadable after 20 values or so, then completely flattening out this argument
- There is a small overhead as previous value storage, then actual point computation ; the overhead is so minimal that it's not in itself a no-go reason (overhead also present on the `Save` computation)

Then, what would be your pro/cons for ref points?
_(Message c/p from Gitter pm.)_

**Comment by Daynvheur** [comment]
Yay @http://localhost:1234/?complexity=2&pt=794;201&pt=572;423&pt=568;112&pt=862;428&pt=572;421 versus http://localhost:1234/?fftsize=8&range=2&complexity=7&circles=true&pt=800;200&pt=600;400&pt=600;100&pt=900;400&pt=600;400!
imageimage

Waiting for discussion about references points, anything else looks like ready to merge.

File: src/index.ts [hunk 1]
```diff
@@ -7,24 +7,104 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
 let unclosedLength = 0;
 let unclosedPath = new Path2D();
 
-const fftSize = 4096;
-const fft = new FFT(fftSize);
-const input = fft.createComplexArray() as number[];
-const output = fft.createComplexArray() as number[];
+let fftSize = 4096;
+let fft = new FFT(fftSize);
```
Comment (Daynvheur) [inline src/index.ts:11]: Those fftsize actions are a perfect fit for initControls.

File: src/index.ts [hunk 2]
```diff
@@ -7,24 +7,104 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+                            complexityCircles.checked = Boolean(v);
+                            break;
+
+                        case (k === 'complexity'):
+                            complexityNumber.value = v;
+                            break;
+
+                        case (k === 'fftsize'):
+                            fftSize = Number(v);
+                            fft = new FFT(fftSize);
```
Comment (Daynvheur) [inline src/index.ts]: Those fftsize actions are a perfect fit for initControls.
(@Daynvheur to myself)

**Comment by jnm2** [comment]
> Then, what would be your pro/cons for ref points?

You bring up a good use case for leaving them absolute which I wasn't thinking about: you can use it as a quick and dirty way to edit the path which there is currently no other way to do. I think this is a strong reason to keep it absolute.

The only reason to introduce encoding such as differential encoding would be to shrink it, but I agree that it is not a very important goal compared to the ability to edit. I'm happy with your approach.

File: src/index.ts (hunk 1 above)
Comment (Daynvheur) [reply src/index.ts:11]: Those are perfect where they are.

File: src/index.ts (hunk 2 above)
Comment (Daynvheur) [reply src/index.ts]: Those are perfect where they are.

**Comment by Daynvheur** [comment]
If you want to have the differential encoding, you can add a special parameter key (preferably before the point list...) that could toggle a boolean starting the diff `addPoint` operations.

Otherwise, the PR is ready to merge. :)

**Comment by jnm2** [comment]
There is one problem I found. The first point you draw is no longer the start of the path when the slider is all the way to the left.

starts at last point

File: src/index.ts [hunk 3]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+
+                        default:
+                            break;
+                    }
+                }
+            })
+}
+
+function setLocation() {
+    if (points.length > 0 && 'URLSearchParams' in window) {
```
Comment (jnm2) [inline src/index.ts]: If we intend to support browsers where `'URLSearchParams' in window` is false, we should move this check to control the visibility of the button so that we don't show a button that does nothing in those browsers.

File: src/index.ts [hunk 4]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+        let pointsString: string = '';
+        points.forEach(pt => pointsString += '&pt=' + pt.x + ';' + pt.y);
+
+        var newRelativePathQuery = window.location.pathname + '?' + 'range=' + parameter + '&' + 'complexity=' + complexity + '&' + 'circles=' + Number(circles) + pointsString;
```
Comment (jnm2) [inline src/index.ts]: Why not `'&complexity='` like the `&pt=` above?

Comment (jnm2) [inline src/index.ts]: What do you think of using `&circles` with no `=`? True/false would be encoded based on whether the parameter is present or not present.

File: src/index.ts [hunk 5]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+                            break;
+                    }
+                }
+            })
+}
+
+function setLocation() {
+    if (points.length > 0 && 'URLSearchParams' in window) {
+        let pointsString: string = '';
+        points.forEach(pt => pointsString += '&pt=' + pt.x + ';' + pt.y);
```
Comment (jnm2) [inline src/index.ts]: Optionally, consider using template strings: `` `&pt=${pt.x};${pt.y}` ``

File: src/index.ts [hunk 6]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
 function updateCanvasSize() {
     canvas.width = window.devicePixelRatio * canvas.clientWidth;
     canvas.height = window.devicePixelRatio * canvas.clientHeight;
 }
 
+function loadLocation() { //Inspiration from https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript/21152762#21152762 (qd's not stored)
+    if (window.location.search)
+        window.location.search.substr(1).split(`&`)
+            .forEach(item => {
+                let [k, v] = item.split(`=`);
```
Comment (jnm2) [inline src/index.ts:46]: Use of backticks (also two lines above) doesn't match the use of single quotes in the rest of the method.

File: src/index.ts [hunk 7]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+};
+
 function updateCanvasSize() {
     canvas.width = window.devicePixelRatio * canvas.clientWidth;
     canvas.height = window.devicePixelRatio * canvas.clientHeight;
 }
 
+function loadLocation() { //Inspiration from https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript/21152762#21152762 (qd's not stored)
+    if (window.location.search)
+        window.location.search.substr(1).split(`&`)
```
Comment (jnm2) [inline src/index.ts]: Would this work?
```suggestion
    window.location.search?.substr(1).split(`&`)
```

File: src/index.ts [hunk 8]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+            .forEach(item => {
+                let [k, v] = item.split(`=`);
+                if (v != null) { //Restriction to valued keys
+                    v = v && decodeURIComponent(v);
```
Comment (jnm2) [inline src/index.ts]: What do you think about using `https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams` to do less manual work?

https://caniuse.com/urlsearchparams

File: src/index.ts [hunk 9]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
 let unclosedPath = new Path2D();
 
-const fftSize = 4096;
-const fft = new FFT(fftSize);
-const input = fft.createComplexArray() as number[];
-const output = fft.createComplexArray() as number[];
+let fftSize = 4096;
+let fft = new FFT(fftSize);
+let input = fft.createComplexArray() as number[];
+let output = fft.createComplexArray() as number[];
```
Comment (jnm2) [inline src/index.ts:13]: When a set of related variables become mutable, I would eventually like to encapsulate them. I don't have a fully-formed thought on this yet though. One goal will be to reduce the duplication that is being introduced between what these lines are doing and what `loadLocation` does.

File: src/index.ts [hunk 10]
```diff
@@ -7,24 +7,103 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+
+                        case (k === 'fftsize'):
+                            fftSize = Number(v);
+                            fft = new FFT(fftSize);
+                            input = fft.createComplexArray() as number[];
+                            output = fft.createComplexArray() as number[];
+                            break;
+
+                        default:
+                            break;
```
Comment (jnm2) [inline src/index.ts]: I think this can be removed.

**Review by jnm2 — CHANGES_REQUESTED** [review]
I like these new abilities.

**Comment by jnm2** [comment]
I discovered that neither ESLint nor TypeScript errors were showing up as a CI build failure here, so I fixed that.
You can run `yarn eslint . --fix` to fix almost all of these: https://github.com/jnm2/epic/pull/4/checks?check_run_id=1434736378

I was messing around for a while. I tried to set something up that would show the errors inline in the PR using GitHub Checks annotations, but https://github.com/marketplace/actions/lint-action doesn't yet support running on `pull_request_target` triggers which is necessary because GitHub prevents forks from accessing upstream repo tokens that are required in order to create the inline annotations. I'm out of time, so I just gave up on inline annotations for now.
So for now, you'll have to click 'Details' on the PR build failure status message at the bottom of the PR to read about it, or run `yarn validate` if you want to do the checks locally.

File: src/index.ts (hunk 3 above)
Comment (Daynvheur) [reply src/index.ts]: Sure. We might aswell find another solution that works on all (oldest browers), not relying on `URLSearchParams`.

File: src/index.ts (hunk 4 above)
Comment (Daynvheur) [reply src/index.ts]: Ah, why is '&' not glued? So that I can change parameter's order without having to remember which one is the first one; `pointsString` is special, as I didn't special-cased the start/end point(s), nor wanted to get rid of an extra `&`. But yeah, `&complexity` would have been valid.
If that's a strong request, I may change that.

Comment (Daynvheur) [reply src/index.ts]: Hmm, I like the idea of expressing the intent; if I cut the apple in half, the default value for `circles` (without equal sign) would be `true`. I'll add this.

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: I couldn't make the C# `$"{pt.x};{pt.y}"` work, and didn't looked very far. I'll take the `'&pt=${pt.x};${pt.y}'` syntax.

File: src/index.ts (hunk 8 above)
Comment (Daynvheur) [reply src/index.ts]: Because there was no IE11 support (even if I'm not on IE11 myself), and [that](https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript/21152762#21152762) looked like a not very convoluted way to achieve parameters reading.

File: src/index.ts (hunk 10 above)
Comment (Daynvheur) [reply src/index.ts]: Indeed. I'll do it.

Comment (Daynvheur) [reply src/index.ts]: 3ad8c99

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: `?range=2&complexity=2&circles=0&pt=${pt.x};${pt.y}&pt=${pt.x};${pt.y}&pt=${pt.x};${pt.y}&pt=${pt.x};${pt.y}&pt=${pt.x};${pt.y}&pt=${pt.x};${pt.y}`
There's something more to investigate.

File: src/index.ts (hunk 4 above)
Comment (Daynvheur) [reply src/index.ts]: 20b1c85

**Comment by Daynvheur** [comment]
I'm reviewing lint errors now.

File: src/index.ts (hunk 3 above)
Comment (Daynvheur) [reply src/index.ts]: 7a568d0 removed the string check.

**Comment by Daynvheur** [comment]
>   51:21  error  Unexpected lexical declaration in case block  no-case-declarations
  56:33  error  Unexpected lexical declaration in case block  no-case-declarations

Could you permit them?

**Comment by jnm2** [comment]
Sure, I guess I'm okay with that. The rule is on the ESLint recommended list though: https://eslint.org/docs/rules/no-case-declarations

This is something you can fix. Either wrap the case block in `{`...`}` like the link above shows, or insert `"no-case-declarations": "off",` in `.eslintrc` in the repo root.

File: src/index.ts [hunk 11]
```diff
@@ -60,44 +145,23 @@ document.getElementById('clear-button')!.onclick = function() {
...
         unclosedLength += segmentLength;
-
-        const addedPoint = { x, y, segmentLength };
-        points.splice(points.length - 1, 0, addedPoint);
-
-        const startAndEndPoint = points[points.length - 1];
-        startAndEndPoint.segmentLength = magnitude(startAndEndPoint.x - x, startAndEndPoint.y - y);
+        points[points.length - 1] = { x, y, segmentLength };
+        const firstPoint = points[0];
+        points.push({ x: firstPoint.x, y: firstPoint.y, segmentLength: magnitude(firstPoint.x - x, firstPoint.y - y) });
```
Comment (jnm2) [inline src/index.ts:164]: Why these changes and the ones in `samplePathIntoInput` (https://github.com/jnm2/epic/pull/4/commits/c76407139c1d06828882d075425ab146e9c40c0c)? I think it is undoing https://github.com/jnm2/epic/commit/370c1ba7395ece25708e3f54483be247599131fc.

File: src/index.ts (hunk 8 above)
Comment (jnm2) [reply src/index.ts]: Okay. I usually ignore IE completely in personal projects because it's less fun, but I'm fine with this.

File: src/index.ts (hunk 4 above)
Comment (jnm2) [reply src/index.ts]: Thanks, just curious. I'm fine with this.

File: src/index.ts (hunk 5 above)
Comment (jnm2) [reply src/index.ts]: @Daynvheur I think you missed the fact that JavaScript template strings use backticks rather than single quotes.

`` `&pt=${pt.x};${pt.y}` ``, not 
`'&pt=${pt.x};${pt.y}'`

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: src/index.ts [hunk 12]
```diff
@@ -7,24 +7,109 @@ const points = new Array<{ x: number, y: number, segmentLength: number }>();
...
+                case 'circles':
+                    complexityCircles.checked = true;
+                    break;
+
+                default:
+                    const [k, v] = item.split('=');
+                    if (v !== null) { // Restriction to valued keys
+                        const w = v && decodeURIComponent(v);
+                        switch (true) {
+                            case (k === 'pt'):
```
Comment (jnm2) [inline src/index.ts:55]: Just curious, is there an advantage to this over `switch (k)`?

Comment (jnm2) [inline src/index.ts]: Just curious, is there an advantage to this over `switch (k)`?

Comment (Daynvheur) [reply src/index.ts]: Of course: if there is the need for a `k AND w`-based case.
...
Okay, this is likely not gonna happen anytime soon, I'll remove that. :)

Comment (Daynvheur) [reply src/index.ts]: 065c652

Comment (jnm2) [reply src/index.ts]: I would move the AND part inside the case then, if I follow what you're saying.

File: src/index.ts [hunk 13]
```diff
@@ -60,44 +148,23 @@ document.getElementById('clear-button')!.onclick = function() {
...
         unclosedLength += segmentLength;
-
-        const addedPoint = { x, y, segmentLength };
-        points.splice(points.length - 1, 0, addedPoint);
-
-        const startAndEndPoint = points[points.length - 1];
-        startAndEndPoint.segmentLength = magnitude(startAndEndPoint.x - x, startAndEndPoint.y - y);
+        points[points.length - 1] = { x, y, segmentLength };
+        const firstPoint = points[0];
+        points.push({ x: firstPoint.x, y: firstPoint.y, segmentLength: magnitude(firstPoint.x - x, firstPoint.y - y) });
```
Comment (jnm2) [inline src/index.ts]: Could you do this without changing how this works? This is reverting newer code that I think I prefer.
It would then not be necessary to skip any points. If you prefer, you could move the last point to become first when you save, but there is really an equivalence since it's a loop.

**Review by jnm2 — CHANGES_REQUESTED** [review]
I'm sorry, thought I left this comment in the previous review round but I don't see it.

**Comment by jnm2** [comment]
@Daynvheur I edited the PR body to say "Also fixes #5." This is recognized by GitHub and creates a special relationship between this PR and issue #5.

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: Oh, I've always thought backticks were `'` duplicates.
Happy to see one use-case, I'll edit that.

File: src/index.ts (hunk 9 above)
Comment (Daynvheur) [reply src/index.ts:13]: (Leaving a comment after conversation resolution)
I thought those could go inside the `initControls` function; that raised errors, I then decided they were perfectly fine as-is.
But one day, they might go somewhere better thought. :)

File: src/index.ts (hunk 13 above)
Comment (Daynvheur) [reply src/index.ts]: I did some wrong copy-past, didn't I?
I think that was part of the d025c9b reapplication…
Edit : I misunderstood the commit, `startAndEndPoint` was the code actually to conserve.

Comment (Daynvheur) [reply src/index.ts]: 138ef2b - sorry for the mispasted code.

File: src/index.ts (hunk 5 above)
Comment (Daynvheur) [reply src/index.ts]: 7d8461e

File: src/index.ts (hunk 13 above)
Comment (jnm2) [reply src/index.ts]: No worries! Thanks for all the work you're doing!

**Comment by jnm2** [comment]
@Daynvheur Thanks again for the cool additions!

File: src/index.ts [hunk 14]
```diff
@@ -97,7 +166,7 @@ function addPoint(x: number, y: number) {
         points.splice(points.length - 1, 0, addedPoint);
 
         const startAndEndPoint = points[points.length - 1];
-        startAndEndPoint.segmentLength = magnitude(startAndEndPoint.x - x, startAndEndPoint.y - y);
+        startAndEndPoint.segmentLength = magnitude(startAndEndPoint.x - x, startAndEndPoint.y - y);
```
Comment (jnm2) [inline src/index.ts:169]: I think the way to fix these issues with line endings is to have a .gitattributes in the repo to enable autocrlf.

Comment (jnm2) [reply src/index.ts:169]: Resolved in https://github.com/jnm2/epic/commit/921618ed677e0f15069769ca39eb14fb6f90e495 with new .gitattributes file and `git add . --renormalize`.
